File: erlang/etudes/ch4/powers.erl
-module(powers).
-export([raise/2,nth_root/2]).

raise(_, 0) -> 1;
raise(X, N) when N < 0 -> 1 / raise(X, 0 - N);
raise(X, N) -> raise(X, N, 1).

raise(X, 1, Accumulator) -> X * Accumulator;
raise(X, N, Accumulator) -> raise(X, N - 1, X * Accumulator).

nth_root(X, N) -> nth_root(X, N, X / 2.0).
nth_root(X, N, A) ->
    io:format("Current guess is ~w~n",[A]),
    F = raise(A, N) - X,
    Fprime = N * raise(A, N - 1),
    Next =  A - F / Fprime,
    Change = Next - A,

    if 
        abs(Change) < 1.0e-8 -> Next;
        true -> nth_root(X, N, Next)
    end.

File: elixir/etudes/ch04/powers.ex
defmodule Powers do
  import Kernel, except: [raise: 2, raise: 3]

  def raise(x, n) when n < 0, do: 1 / raise(x, abs(n))
  def raise(x, n), do: raise(x, n, 1)

  defp raise(_x, n, acc) when n == 0, do: acc
  defp raise(x, n, acc), do: raise(x, n - 1, x * acc)

  def nth_root(x, n), do: nth_root(x, n, x / 2.0)
  defp nth_root(x, n, a) do
    IO.puts("Current guess is #{a}")
    f = raise(a, n) - x
    fprime = n * raise(a, n - 1)
    next =  a - f / fprime
    change = next - a

    if abs(change) < 1.0e-8 do
      next
    else
      nth_root(x, n, next)
    end
  end


end
